File: prak2/cpp/elgamal.cpp
#include "numbInDegree.h"
#include "evklidAlg.h"
#include "gen_p.h"
#include "elgamal.h"

#include <iostream>
#include <fstream>
#include <vector>
#include <random>
#include <clocale>
#include <codecvt>
#include <locale>

using namespace std;

// Устанавливаем локаль для работы с UTF-8
void setUTF8Locale() {
    setlocale(LC_ALL, "ru_RU.UTF-8");
    std::locale::global(std::locale("ru_RU.UTF-8"));
}

int search_g(int p) {
    vector<int> fact;
    int phi = p - 1, n = phi;
    for (int i = 2; i * i <= n; ++i) {
        if (n % i == 0) {
            fact.push_back(i);
            while (n % i == 0)
                n /= i;
        }
    }
    if (n > 1) fact.push_back(n);
    for (int res = 2; res <= p; ++res) {
        bool ok = true;
        for (size_t i = 0; i < fact.size() && ok; ++i) 
            ok &= bin_degre(res, phi / fact[i], p) != 1;
        if (ok) return res;
    }
    return -1;
}

string numbToChar(int num) {
    wstring_convert<codecvt_utf8<char32_t>, char32_t> converter; //Инициализация конвертера.
    // Преобразуем число в один Unicode-символ
    char32_t unicode_char = static_cast<char32_t>(num + 0x20); // Сдвиг на 0x20
    return converter.to_bytes(&unicode_char, &unicode_char + 1); //возвращаем символ конвернтированный в utf-8. &unicode_char, &unicode_char + 1 указывают на начало и конец "строки" символа в utf-8 т.е. на 1 символ в этой кодировке 
}

int charToNumb(const string& utf8str) {
    wstring_convert<codecvt_utf8<char32_t>, char32_t> converter;
    // Конвертируем строку UTF-8 в UTF-32 (ожидаем ровно один символ)
    u32string utf32 = converter.from_bytes(utf8str); //Конвернтируем в  utf-32 u32string
    return static_cast<int>(utf32[0]) - 0x20; // Обратный сдвиг
}

int elgamal_encription() {
    setUTF8Locale();
    wifstream source_text("sourse_text.txt"); //используем широкие потоки для работы с utf-8
    int p = gen_p(15, 8, 8); //генерируем 15 битный ключ
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> random_num(2, p - 1);

    int g = search_g(p);
    int x = random_num(gen);
    int y = bin_degre(g, x, p);
    int k = random_num(gen);

    ofstream cipherText("cipher.txt");

    // Записываем публичный ключ
    cipherText << numbToChar(y) << " " << numbToChar(g) << " " << numbToChar(p) << " ";

    // Шифруем текст
    int a = bin_degre(g, k, p);
    wchar_t wc;
    source_text >> noskipws; //отключаем пропуск пробелов и символов конца строк
    while (source_text >> wc) {
        int M = static_cast<int>(wc);
        int b = bin_degre(y, k, p) * M % p;
        
        // Записываем пару (a, b)
        cipherText << numbToChar(a) << " " << numbToChar(b) << " ";
    }

    source_text.close();
    cipherText.close();
    cout << "Файл успешно зашифрован" << endl;
    return x;
}

void elgamal_decription(int x) {
    setUTF8Locale();
    
    ifstream cipherText("cipher.txt");

    tuple<int, int, int> pub_key;
    string y_str, g_str, p_str;
    cipherText >> y_str >> g_str >> p_str;

    get<0>(pub_key) = charToNumb(y_str);
    get<1>(pub_key) = charToNumb(g_str);
    get<2>(pub_key) = charToNumb(p_str);

    wofstream completeText("compliteText.txt");
    completeText.imbue(std::locale("ru_RU.UTF-8"));

    string a_str, b_str;
    while (cipherText >> a_str >> b_str) {
        int a = charToNumb(a_str);
        int b = charToNumb(b_str);

        int a_x = bin_degre(a, x, get<2>(pub_key));
        int M = evklid_algoritm_mul(a_x, get<2>(pub_key)) * b % get<2>(pub_key);
        completeText << static_cast<wchar_t>(M);
    }

    cipherText.close();
    completeText.close();
    cout << "Файл успешно расшифрован" << endl;
}

void brutforse() {
    cout << "Выберите предполагаемый язык текста: 1 - русский, 2 - английский" << endl;
    int language;
    cin >> language;
    if (cin.fail() || language < 1 || language > 2) {
        cerr << "Ошибка выбора языка!" << endl;
        return;
    }

    setUTF8Locale();

    ifstream cipherText("cipher.txt");

    tuple<int, int, int> pub_key;
    string y_str, g_str, p_str;
    cipherText >> y_str >> g_str >> p_str;

    get<0>(pub_key) = charToNumb(y_str);
    get<1>(pub_key) = charToNumb(g_str);
    get<2>(pub_key) = charToNumb(p_str);

    for (int x = 1; x < 10000; x++) {
        cipherText.clear();
        cipherText.seekg(0);
        
        // Пропускаем публичный ключ
        string tmp;
        for (int i = 0; i < 3; i++) cipherText >> tmp;

        bool success = true;
        wofstream completeText("compliteText.txt");

        string a_str, b_str;
        while (cipherText >> a_str >> b_str) {
            int a = charToNumb(a_str);
            int b = charToNumb(b_str);

            int a_x = bin_degre(a, x, get<2>(pub_key));
            int M = (evklid_algoritm_mul(a_x, get<2>(pub_key)) * b) % get<2>(pub_key);
            wchar_t wc = static_cast<wchar_t>(M);

            // Проверка на валидность символа
            if (language == 2) { // Английский
                if (!(wc == L'\n' || (wc >= L' ' && wc <= L'~'))) {
                    success = false;
                    break;
                }
            } else { // Русский
                if (!(wc == L'\n' || (wc >= L'А' && wc <= L'я') || wc == L'Ё' || wc == L'ё' || 
                     (wc >= L' ' && wc <= L'`') || (wc >= L'{' && wc <= L'~'))) {
                    success = false;
                    break;
                }
            }
            completeText << wc;
        }

        if (success) {
            cout << "Успех! Закрытый ключ x = " << x << endl;
            cout << "Расшифрованный текст сохранен в compliteText.txt" << endl;
            completeText.close();
            cipherText.close();
            return;
        }
        completeText.close();
    }

    cipherText.close();
    cout << "Взлом не удался." << endl;
}

enum choiceType {
    DECR = 1,
    BRUT
};

void cipherElGam() {
    int x = elgamal_encription();
    if (x == -1) return;

    cout << "Введите 1 для дешифрования, 2 для эмуляции атаки" << endl;
    string usCin;
    try {
        cin >> usCin;
        if (usCin[0] < '1' || usCin[0] > '2' || usCin.size() != 1) {
            throw usCin;
        }
    }
    catch (const string& error) {
        cerr << "Ошибка, вы ввели: " << error << ", а можно вводить только 1 или 2" << endl;
        exit(0);
    }

    auto choice = static_cast<choiceType>(usCin[0] - '0');
    switch (choice) {
        case DECR:
            elgamal_decription(x);
            break;
        case BRUT:
            brutforse();
            break;
        default:
            break;
    }
}

File: prak2/go/elgamal.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"unicode/utf8"
)

// Поиск первообразного корня
func search_g(p int) int {
	var fact []int
	phi := p - 1
	n := phi

	for i := 2; i*i <= n; i++ {
		if n%i == 0 {
			fact = append(fact, i)
			for n%i == 0 {
				n /= i
			}
		}
	}
	if n > 1 {
		fact = append(fact, n)
	}

	for res := 2; res <= p; res++ {
		ok := true
		for i := 0; i < len(fact) && ok; i++ {
			if binDegre(res, phi/fact[i], p) == 1 {
				ok = false
			}
		}
		if ok {
			return res
		}
	}
	return -1
}

// преобразование числа в символ utf-8
func numbToChar(num int) string {
	return string(rune(num + 0x20))
}

// преобразование символа utf-8 в число
func charToNumb(utf8str string) int {
	r, _ := utf8.DecodeRuneInString(utf8str)
	return int(r) - 0x20
}

// Шифрование
func elgamal_encription() int {
	sourceFile, err := os.Open("sourse_text.txt")
	if err != nil {
		fmt.Println("Error opening source file:", err)
		return -1
	}
	defer sourceFile.Close()

	p := gen_p(15, 8, 8)
	randomNum := rand.Intn(p-2) + 2

	g := search_g(p)
	x := randomNum
	y := binDegre(g, x, p)
	k := randomNum

	cipherFile, err := os.Create("cipher.txt")
	if err != nil {
		fmt.Println("Error creating cipher file:", err)
		return -1
	}
	defer cipherFile.Close()

	cipherFile.WriteString(numbToChar(y) + " " + numbToChar(g) + " " + numbToChar(p) + " ")

	a := binDegre(g, k, p)
	scanner := bufio.NewScanner(sourceFile) // Создаём сканер для чтения данных из файла
	scanner.Split(bufio.ScanRunes)          // Раразбиваем поток байтов на отдельные юникод символы

	for scanner.Scan() {
		char := scanner.Text()
		r, _ := utf8.DecodeRuneInString(char)
		M := int(r)
		b := binDegre(y, k, p) * M % p

		cipherFile.WriteString(numbToChar(a) + " " + numbToChar(b) + " ")
	}

	fmt.Println("Файл успешно зашифрован")
	return x
}

// Дешифрование
func elgamal_decription(x int) {
	cipherFile, _ := os.Open("cipher.txt")
	defer cipherFile.Close() //закрываем файл после выполнение функции с помощью ключевого слова defer

	scanner := bufio.NewScanner(cipherFile)
	scanner.Split(bufio.ScanWords)

	//считываем публичный ключ
	scanner.Scan()
	scanner.Scan()
	scanner.Scan()
	p_str := scanner.Text()
	p := charToNumb(p_str)

	completeFile, _ := os.Create("compliteText.txt")
	defer completeFile.Close() //закрываем файл после выполнение функции с помощью ключевого слова defer
	for scanner.Scan() {
		a_str := scanner.Text()
		scanner.Scan()
		b_str := scanner.Text()

		a := charToNumb(a_str)
		b := charToNumb(b_str)

		a_x := binDegre(a, x, p)
		M := evklidAlgoritmMul(a_x, p, 0) * b % p
		completeFile.WriteString(string(rune(M)))
	}

	fmt.Println("Файл успешно расшифрован")
}

// Эмуляция атаки методом брутфорс
func brutforse() {
	fmt.Println("Выберите предполагаемый язык текста: 1 - русский, 2 - английский")
	var language int
	_, err := fmt.Scan(&language)
	if err != nil || language < 1 || language > 2 {
		fmt.Println("Ошибка выбора языка!")
		return
	}

	cipherFile, _ := os.Open("cipher.txt")
	defer cipherFile.Close()

	scanner := bufio.NewScanner(cipherFile)
	scanner.Split(bufio.ScanWords)

	scanner.Scan()
	scanner.Scan()
	scanner.Scan()
	p_str := scanner.Text()
	p := charToNumb(p_str)

	for x := 1; x < 10000; x++ {
		cipherFile.Seek(0, 0) // Сбрасываем позицию сканера (ставим на начало файла)
		scanner = bufio.NewScanner(cipherFile)
		scanner.Split(bufio.ScanWords)

		// пропуск публичного ключа
		for i := 0; i < 3; i++ {
			scanner.Scan()
		}

		success := true
		completeFile, _ := os.Create("compliteText.txt")

		for scanner.Scan() {
			a_str := scanner.Text()
			scanner.Scan()
			b_str := scanner.Text()

			a := charToNumb(a_str)
			b := charToNumb(b_str)

			a_x := binDegre(a, x, p)
			M := (evklidAlgoritmMul(a_x, p, 0) * b) % p
			wc := rune(M)

			if language == 2 { // Английский
				if !(wc == '\n' || (wc >= ' ' && wc <= '~')) {
					success = false
					break
				}
			} else { // Русский
				if !(wc == '\n' || (wc >= 'А' && wc <= 'я') || wc == 'Ё' || wc == 'ё' ||
					(wc >= ' ' && wc <= '`') || (wc >= '{' && wc <= '~')) {
					success = false
					break
				}
			}
			completeFile.WriteString(string(wc))
		}

		completeFile.Close()

		if success {
			fmt.Printf("Успех! Закрытый ключ x = %d\n", x)
			return
		}
	}

	fmt.Println("Взлом не удался.")
}

type choiceType int

const (
	DECR choiceType = iota + 1
	BRUT
)

// Основная функция программы
func cipherElGam() {
	x := elgamal_encription()
	if x == -1 {
		return
	}

	fmt.Println("Введите 1 для дешифрования, 2 для эмуляции атаки")
	var usCin string
	_, err := fmt.Scan(&usCin)
	if err != nil || len(usCin) != 1 || usCin[0] < '1' || usCin[0] > '2' {
		fmt.Printf("Ошибка, вы ввели: %s, а можно вводить только 1 или 2\n", usCin)
		os.Exit(0)
	}

	choice, _ := strconv.Atoi(usCin)
	switch choiceType(choice) {
	case DECR:
		elgamal_decription(x)
	case BRUT:
		brutforse()
	default:
		break
	}
}
